src/python/pants/backend/python/tasks/python_execution_task_base.py

```diff
@@ -117,37 +60,22 @@ def create_pex(self, pex_info=None):
         target_set_id = 'no_targets'
 
       interpreter = self.context.products.get_data(PythonInterpreter)
-      path = os.path.join(self.workdir, str(interpreter.identity), target_set_id)
-      extra_pex_paths = None
+      path = os.path.realpath(os.path.join(self.workdir, str(interpreter.identity), target_set_id))
 
       # Note that we check for the existence of the directory, instead of for invalid_vts,
       # to cover the empty case.
       if not os.path.isdir(path):
         source_pexes = self.context.products.get_data(GatherSources.PythonSources).all()
         requirements_pex = self.context.products.get_data(ResolveRequirements.REQUIREMENTS_PEX)
         pexes = [requirements_pex] + source_pexes
-
         if self.extra_requirements():
-          extra_reqs = [PythonRequirement(req_str) for req_str in self.extra_requirements()]
-          addr = Address.parse('{}_extra_reqs'.format(self.__class__.__name__))
-          self.context.build_graph.inject_synthetic_target(
-            addr, PythonRequirementLibrary, requirements=extra_reqs)
+          extra_requirements_pex = self.resolve_requirement_strings(
+            interpreter, self.extra_requirements())
           # Add the extra requirements first, so they take precedence over any colliding version
           # in the target set's dependency closure.
-          pexes = [self.resolve_requirements([self.context.build_graph.get_target(addr)])] + pexes
-
-        extra_pex_paths = [pex.path() for pex in pexes if pex]
-
-        if extra_pex_paths:
-          pex_info.merge_pex_path(':'.join(extra_pex_paths))
-
-        with safe_concurrent_creation(path) as safe_path:
-          builder = PEXBuilder(safe_path, interpreter, pex_info=pex_info)
-          # Add target interpreter compatibilities to pex info.
-          for rt in relevant_targets:
-            if has_python_sources(rt):
-              for constraint in rt.compatibility:
-                builder.add_interpreter_constraint(constraint)
-          builder.freeze()
+          pexes = [extra_requirements_pex] + pexes
+        constraints = {constraint for rt in relevant_targets if has_python_sources(rt)
```

Comment:
Might be helpful to name this more explicitly to `interpreter_constraints` since this was previously explained by the `builder.add_interpreter_constraint(constraint)` on the following line, and now this info is not immediately inferable from the file context alone. However, one could argue that experienced devs will get this and this is more of a concern for those who are not super-familiar with Python targets, so just a suggestion.